File: pi_agent/lib/src/agent.dart
/// Agent class - stateful wrapper around the agent loop.
///
/// See full implementation in Phase 4 tasks.
library;

import 'dart:async';

import 'types.dart';

/// Callback type for agent event listeners.
typedef AgentEventListener = FutureOr<void> Function(AgentEvent event);

/// Options for constructing an Agent.
class AgentOptions {
  /// Initial model.
  final Model? model;

  /// Initial system prompt.
  final String? systemPrompt;

  /// Initial thinking level.
  final ThinkingLevel? thinkingLevel;

  /// Initial tools.
  final List<AgentTool<dynamic, dynamic>>? tools;

  /// Initial messages.
  final List<AgentMessage>? messages;

  /// Session identifier.
  final String? sessionId;

  /// Custom stream function (default: uses LLM client).
  final Stream<AgentEvent> Function(
      AgentContext context, AgentLoopConfig config)? streamFn;

  /// Custom message converter.
  final Future<List<Map<String, dynamic>>> Function(List<AgentMessage>)?
      convertToLlm;

  /// Custom context transformer.
  final Future<List<AgentMessage>> Function(List<AgentMessage>,
      {bool Function()? isAborted})? transformContext;

  /// API key resolver.
  final Future<String?> Function(String provider)? getApiKey;

  /// Before tool call hook.
  final Future<BeforeToolCallResult?> Function(BeforeToolCallContext,
      {bool Function()? isAborted})? beforeToolCall;

  /// After tool call hook.
  final Future<AfterToolCallResult?> Function(AfterToolCallContext,
      {bool Function()? isAborted})? afterToolCall;

  /// Tool execution mode.
  final ToolExecutionMode? toolExecution;

  /// Steering mode.
  final SteeringMode? steeringMode;

  /// Follow-up mode.
  final FollowUpMode? followUpMode;

  /// Thinking budget tokens per level.
  final Map<ThinkingLevel, int>? thinkingBudgets;

  /// Extra headers for LLM requests.
  final Map<String, String>? extraHeaders;

  /// Creates agent options.
  const AgentOptions({
    this.model,
    this.systemPrompt,
    this.thinkingLevel,
    this.tools,
    this.messages,
    this.sessionId,
    this.streamFn,
    this.convertToLlm,
    this.transformContext,
    this.getApiKey,
    this.beforeToolCall,
    this.afterToolCall,
    this.toolExecution,
    this.steeringMode,
    this.followUpMode,
    this.thinkingBudgets,
    this.extraHeaders,
  });
}

/// Stateful agent managing conversation lifecycle, tool execution, and events.
///
/// The Agent wraps the low-level agent loop with:
/// - Mutable state management (tools, messages, system prompt)
/// - Event subscription system
/// - Steering and follow-up message queues
/// - Abort/cancellation support
class Agent {
  final AgentState _state;
  final String? _sessionId;
  Stream<AgentEvent> Function(AgentContext context, AgentLoopConfig config)?
      _streamFn;
  Future<List<Map<String, dynamic>>> Function(List<AgentMessage>)?
      _convertToLlm;
  Future<List<AgentMessage>> Function(List<AgentMessage>,
      {bool Function()? isAborted})? _transformContext;
  Future<String?> Function(String provider)? _getApiKey;
  Future<BeforeToolCallResult?> Function(BeforeToolCallContext,
      {bool Function()? isAborted})? _beforeToolCall;
  Future<AfterToolCallResult?> Function(AfterToolCallContext,
      {bool Function()? isAborted})? _afterToolCall;
  ToolExecutionMode _toolExecution;
  // ignore: unused_field
  SteeringMode _steeringMode;
  // ignore: unused_field
  FollowUpMode _followUpMode;
  // ignore: unused_field
  final Map<ThinkingLevel, int>? _thinkingBudgets;
  final Map<String, String>? _extraHeaders;

  final List<AgentEventListener> _listeners = [];
  final List<AgentMessage> _steeringQueue = [];
  final List<AgentMessage> _followUpQueue = [];
  Completer<void>? _abortCompleter;
  bool _isRunning = false;
  Future<void>? _activeRun;

  /// Creates an agent.
  ///
  /// Optionally accepts [AgentOptions] or individual parameters.
  Agent({
    AgentOptions? options,
    Model? model,
    String? systemPrompt,
    ThinkingLevel? thinkingLevel,
    List<AgentTool<dynamic, dynamic>>? tools,
    List<AgentMessage>? messages,
    String? sessionId,
    Stream<AgentEvent> Function(AgentContext context, AgentLoopConfig config)?
        streamFn,
    Future<List<Map<String, dynamic>>> Function(List<AgentMessage>)?
        convertToLlm,
    Future<List<AgentMessage>> Function(List<AgentMessage>,
            {bool Function()? isAborted})?
        transformContext,
    Future<String?> Function(String provider)? getApiKey,
    Future<BeforeToolCallResult?> Function(BeforeToolCallContext,
            {bool Function()? isAborted})?
        beforeToolCall,
    Future<AfterToolCallResult?> Function(AfterToolCallContext,
            {bool Function()? isAborted})?
        afterToolCall,
    ToolExecutionMode? toolExecution,
    SteeringMode? steeringMode,
    FollowUpMode? followUpMode,
    Map<ThinkingLevel, int>? thinkingBudgets,
    Map<String, String>? extraHeaders,
  })  : _sessionId = options?.sessionId ?? sessionId,
        _streamFn = options?.streamFn ?? streamFn,
        _convertToLlm = options?.convertToLlm ?? convertToLlm,
        _transformContext = options?.transformContext ?? transformContext,
        _getApiKey = options?.getApiKey ?? getApiKey,
        _beforeToolCall = options?.beforeToolCall ?? beforeToolCall,
        _afterToolCall = options?.afterToolCall ?? afterToolCall,
        _toolExecution = options?.toolExecution ??
            toolExecution ??
            ToolExecutionMode.parallel,
        _steeringMode =
            options?.steeringMode ?? steeringMode ?? SteeringMode.oneAtATime,
        _followUpMode =
            options?.followUpMode ?? followUpMode ?? FollowUpMode.oneAtATime,
        _thinkingBudgets = options?.thinkingBudgets ?? thinkingBudgets,
        _extraHeaders = options?.extraHeaders ?? extraHeaders,
        _state = AgentState(
          systemPrompt: options?.systemPrompt ?? systemPrompt ?? '',
          model: options?.model ??
              model ??
              const Model(
                  provider: 'openai', modelId: 'gpt-4o', contextWindow: 128000),
          thinkingLevel:
              options?.thinkingLevel ?? thinkingLevel ?? ThinkingLevel.off,
          tools: options?.tools ?? tools ?? [],
          messages: options?.messages ?? messages ?? [],
        );

  /// Current agent state.
  AgentState get state => _state;

  /// Whether the agent is currently running.
  bool get isRunning => _isRunning;

  /// Session identifier.
  String? get sessionId => _sessionId;

  /// Sets the message converter.
  set convertToLlm(
          Future<List<Map<String, dynamic>>> Function(List<AgentMessage>)?
              fn) =>
      _convertToLlm = fn;

  /// Sets the context transformer.
  set transformContext(
          Future<List<AgentMessage>> Function(List<AgentMessage>,
                  {bool Function()? isAborted})?
              fn) =>
      _transformContext = fn;

  /// Sets the stream function.
  set streamFn(
          Stream<AgentEvent> Function(
                  AgentContext context, AgentLoopConfig config)?
              fn) =>
      _streamFn = fn;

  /// Sets the API key resolver.
  set getApiKey(Future<String?> Function(String provider)? fn) =>
      _getApiKey = fn;

  /// Sets the before-tool-call hook.
  set beforeToolCall(
          Future<BeforeToolCallResult?> Function(BeforeToolCallContext,
                  {bool Function()? isAborted})?
              fn) =>
      _beforeToolCall = fn;

  /// Sets the after-tool-call hook.
  set afterToolCall(
          Future<AfterToolCallResult?> Function(AfterToolCallContext,
                  {bool Function()? isAborted})?
              fn) =>
      _afterToolCall = fn;

  /// Sets tool execution mode.
  set toolExecution(ToolExecutionMode mode) => _toolExecution = mode;

  /// Sets steering mode.
  set steeringMode(SteeringMode mode) => _steeringMode = mode;

  /// Sets follow-up mode.
  set followUpMode(FollowUpMode mode) => _followUpMode = mode;

  /// Subscribe to agent events.
  ///
  /// Returns an unsubscribe function. Listeners are called sequentially
  /// and awaited before the agent is considered idle.
  void Function() subscribe(AgentEventListener listener,
      {void Function()? onCancel}) {
    _listeners.add(listener);
    return () => _listeners.remove(listener);
  }

  /// Start a new conversation by sending a prompt.
  ///
  /// [input] can be a `String`, `List<ContentBlock>`, `AgentMessage`, or
  /// `List<AgentMessage>`.
  Future<void> prompt(dynamic input) async {
    if (_isRunning) {
      throw StateError(
          'Agent is already running. Call abort() or waitForIdle() first.');
    }

    final msgs = _inputToMessages(input);
    _state.internalMessages.addAll(msgs);
    await _run();
  }

  /// Continue the conversation from the current transcript.
  ///
  /// The last message must be a user or tool result message.
  Future<void> continue_() async {
    if (_isRunning) {
      throw StateError(
          'Agent is already running. Call abort() or waitForIdle() first.');
    }

    if (_state.internalMessages.isEmpty) {
      throw StateError('Cannot continue: transcript is empty.');
    }

    await _run();
  }

  /// Queue a steering message to be injected during the next turn.
  void steer(AgentMessage message) => _steeringQueue.add(message);

  /// Queue a follow-up message to be injected after the current turn.
  void followUp(AgentMessage message) => _followUpQueue.add(message);

  /// Clear all queued steering messages.
  void clearSteeringQueue() => _steeringQueue.clear();

  /// Clear all queued follow-up messages.
  void clearFollowUpQueue() => _followUpQueue.clear();

  /// Clear all queued messages.
  void clearAllQueues() {
    _steeringQueue.clear();
    _followUpQueue.clear();
  }

  /// Whether there are any queued messages.
  bool hasQueuedMessages() =>
      _steeringQueue.isNotEmpty || _followUpQueue.isNotEmpty;

  /// Abort the current agent run.
  void abort() {
    _abortCompleter?.complete();
  }

  /// Wait for the agent to become idle after a run completes.
  Future<void> waitForIdle() async {
    await _activeRun;
  }

  /// Reset the agent state, clearing transcript and queues.
  void reset() {
    _state.internalMessages.clear();
    _steeringQueue.clear();
    _followUpQueue.clear();
    _state.errorMessage = null;
    _state.isStreaming = false;
    _state.streamingMessage = null;
    _state.internalPendingToolCalls.clear();
  }

  List<AgentMessage> _inputToMessages(dynamic input) {
    if (input is String) return [UserMessage.text(input)];
    if (input is List<ContentBlock>) return [UserMessage(content: input)];
    if (input is AgentMessage) return [input];
    if (input is List<AgentMessage>) return input;
    throw ArgumentError('Unsupported input type: ${input.runtimeType}');
  }

  Future<void> _run() async {
    _isRunning = true;
    _abortCompleter = Completer<void>();
    _state.isStreaming = true;
    _state.errorMessage = null;

    try {
      await _emitEvent(AgentStart(sessionId: _sessionId));

      final config = _buildConfig();
      final context = AgentContext(
        systemPrompt: _state.systemPrompt,
        messages: List.from(_state.internalMessages),
        tools: _state.internalTools.isNotEmpty
            ? List.from(_state.internalTools)
            : null,
      );

      final streamFn = _streamFn ?? _defaultStreamFn;
      final eventStream = streamFn(context, config);

      await for (final event in eventStream) {
        if (_abortCompleter?.isCompleted ?? false) break;
        _handleEvent(event);
        await _emitEvent(event);
      }

      await _emitEvent(AgentEnd(messages: List.from(_state.internalMessages)));
    } catch (e) {
      _state.errorMessage = e.toString();
      await _emitEvent(AgentEnd(messages: List.from(_state.internalMessages)));
    } finally {
      _state.isStreaming = false;
      _state.streamingMessage = null;
      _isRunning = false;
      _abortCompleter = null;
    }
  }

  void _handleEvent(AgentEvent event) {
    switch (event) {
      case MessageStart(message: final msg):
        _state.streamingMessage = msg;
      case MessageUpdate(message: final msg):
        _state.streamingMessage = msg;
      case MessageEnd(message: final msg):
        _state.internalMessages.add(msg);
        _state.streamingMessage = null;
      case ToolExecutionStart(toolCallId: final id):
        _state.internalPendingToolCalls.add(id);
      case ToolExecutionEnd(toolCallId: final id):
        _state.internalPendingToolCalls.remove(id);
      default:
        break;
    }
  }

  Future<void> _emitEvent(AgentEvent event) async {
    for (final listener in _listeners) {
      await listener(event);
    }
  }

  AgentLoopConfig _buildConfig() => AgentLoopConfig(
        model: _state.model,
        toolExecution: _toolExecution,
        convertToLlm: _convertToLlm ?? _defaultConvertToLlm,
        transformContext: _transformContext,
        getApiKey: _getApiKey,
        beforeToolCall: _beforeToolCall,
        afterToolCall: _afterToolCall,
        shouldStopAfterTurn: null,
        getSteeringMessages: _steeringQueue.isNotEmpty
            ? () async => _drainQueue(_steeringQueue)
            : null,
        getFollowUpMessages: _followUpQueue.isNotEmpty
            ? () async => _drainQueue(_followUpQueue)
            : null,
        extraHeaders: _extraHeaders,
      );

  List<AgentMessage> _drainQueue(List<AgentMessage> queue) {
    final msgs = List<AgentMessage>.from(queue);
    queue.clear();
    return msgs;
  }

  Stream<AgentEvent> _defaultStreamFn(
      AgentContext context, AgentLoopConfig config) {
    return Stream.empty();
  }

  Future<List<Map<String, dynamic>>> _defaultConvertToLlm(
      List<AgentMessage> msgs) async {
    return msgs
        .whereType<UserMessage>()
        .map((m) => {
              'role': 'user',
              'content':
                  m.content.whereType<TextBlock>().map((b) => b.text).join(),
            })
        .toList();
  }
}

File: python/src/pi_mono/agent/agent.py
import asyncio
import time
from typing import Any, Callable, Coroutine, List, Union, Set, cast

from pi_mono.ai.types import (
    ImageContent,
    Message,
    Model,
    TextContent,
    ThinkingBudgets,
)
from pi_mono.ai.stream import stream_simple
from pi_mono.utils.abort_signals import AbortSignal, AbortController
from pi_mono.agent.types import (
    AgentContext,
    AgentEvent,
    AgentLoopConfig,
    AgentMessage,
    AgentState,
    AgentTool,
    QueueMode,
    StreamFn,
    ToolExecutionMode,
    Transport,
    BeforeToolCallContext,
    AfterToolCallContext,
)
from pi_mono.agent.agent_loop import (
    run_agent_loop,
    run_agent_loop_continue,
    maybe_await,
)


def default_convert_to_llm(messages: List[AgentMessage]) -> List[Message]:
    return [
        m for m in messages if m.get("role") in ("user", "assistant", "toolResult")
    ]  # type: ignore


EMPTY_USAGE = {
    "input": 0,
    "output": 0,
    "cacheRead": 0,
    "cacheWrite": 0,
    "totalTokens": 0,
    "cost": {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0, "total": 0},
}

DEFAULT_MODEL: Model = {
    "id": "unknown",
    "name": "unknown",
    "api": "unknown",
    "provider": "unknown",
    "baseUrl": "",
    "reasoning": False,
    "input": [],
    "cost": {"input": 0, "output": 0, "cacheRead": 0, "cacheWrite": 0},
    "contextWindow": 0,
    "maxTokens": 0,
}


class MutableAgentState:
    def __init__(self, initial_state: dict[str, Any] | None = None) -> None:
        initial = initial_state or {}
        self.systemPrompt: str = initial.get("systemPrompt", "")
        self.model: Model = initial.get("model", DEFAULT_MODEL)
        self.thinkingLevel: Any = initial.get("thinkingLevel", "off")

        self._tools: List[AgentTool] = list(initial.get("tools", []))
        self._messages: List[AgentMessage] = list(initial.get("messages", []))

        self._is_streaming = False
        self._streaming_message: AgentMessage | None = None
        self._pending_tool_calls: Set[str] = set()
        self._error_message: str | None = None

    @property
    def tools(self) -> List[AgentTool]:
        return self._tools

    @tools.setter
    def tools(self, next_tools: List[AgentTool]) -> None:
        self._tools = list(next_tools)

    @property
    def messages(self) -> List[AgentMessage]:
        return self._messages

    @messages.setter
    def messages(self, next_messages: List[AgentMessage]) -> None:
        self._messages = list(next_messages)

    @property
    def isStreaming(self) -> bool:
        return self._is_streaming

    @property
    def streamingMessage(self) -> AgentMessage | None:
        return self._streaming_message

    @property
    def pendingToolCalls(self) -> Set[str]:
        return self._pending_tool_calls

    @property
    def errorMessage(self) -> str | None:
        return self._error_message


class PendingMessageQueue:
    def __init__(self, mode: QueueMode) -> None:
        self.mode: QueueMode = mode
        self._messages: List[AgentMessage] = []

    def enqueue(self, message: AgentMessage) -> None:
        self._messages.append(message)

    def has_items(self) -> bool:
        return len(self._messages) > 0

    def drain(self) -> List[AgentMessage]:
        if self.mode == "all":
            drained = list(self._messages)
            self._messages.clear()
            return drained

        if not self._messages:
            return []

        first = self._messages.pop(0)
        return [first]

    def clear(self) -> None:
        self._messages.clear()


class Agent:
    def __init__(self, options: dict[str, Any] | None = None) -> None:
        opts = options or {}
        self._state = MutableAgentState(opts.get("initialState"))
        self.listeners: Set[Callable[[AgentEvent, AbortSignal], Any]] = set()

        self.convertToLlm: Callable[[list[AgentMessage]], Union[list[Message], Any]] = (
            opts.get("convertToLlm") or default_convert_to_llm
        )
        self.transformContext: Callable[[list[AgentMessage], AbortSignal | None], Any] | None = (
            opts.get("transformContext")
        )
        self.streamFn: StreamFn = opts.get("streamFn") or stream_simple
        self.getApiKey: Callable[[str], Any] | None = opts.get("getApiKey")
        self.onPayload: Callable[[Any, Model], Any] | None = opts.get("onPayload")
        self.onResponse: Callable[[Any, Model], Any] | None = opts.get("onResponse")
        self.beforeToolCall: Callable[[BeforeToolCallContext, AbortSignal | None], Any] | None = (
            opts.get("beforeToolCall")
        )
        self.afterToolCall: Callable[[AfterToolCallContext, AbortSignal | None], Any] | None = (
            opts.get("afterToolCall")
        )
        self.prepareNextTurn: Callable[[AbortSignal | None], Any] | None = opts.get(
            "prepareNextTurn"
        )

        self.steeringQueue = PendingMessageQueue(opts.get("steeringMode") or "one-at-a-time")
        self.followUpQueue = PendingMessageQueue(opts.get("followUpMode") or "one-at-a-time")

        self.sessionId: str | None = opts.get("sessionId")
        self.thinkingBudgets: ThinkingBudgets | None = opts.get("thinkingBudgets")
        self.transport: Transport = opts.get("transport") or "auto"
        self.maxRetryDelayMs: int | None = opts.get("maxRetryDelayMs")
        self.toolExecution: ToolExecutionMode = opts.get("toolExecution") or "parallel"

        self.active_run: dict[str, Any] | None = None

    def subscribe(self, listener: Callable[[AgentEvent, AbortSignal], Any]) -> Callable[[], None]:
        self.listeners.add(listener)
        return lambda: self.listeners.discard(listener)

    @property
    def state(self) -> AgentState:
        return self._state  # type: ignore

    @property
    def steeringMode(self) -> QueueMode:
        return self.steeringQueue.mode

    @steeringMode.setter
    def steeringMode(self, mode: QueueMode) -> None:
        self.steeringQueue.mode = mode

    @property
    def followUpMode(self) -> QueueMode:
        return self.followUpQueue.mode

    @followUpMode.setter
    def followUpMode(self, mode: QueueMode) -> None:
        self.followUpQueue.mode = mode

    def steer(self, message: AgentMessage) -> None:
        self.steeringQueue.enqueue(message)

    def followUp(self, message: AgentMessage) -> None:
        self.followUpQueue.enqueue(message)

    def clearSteeringQueue(self) -> None:
        self.steeringQueue.clear()

    def clearFollowUpQueue(self) -> None:
        self.followUpQueue.clear()

    def clearAllQueues(self) -> None:
        self.clearSteeringQueue()
        self.clearFollowUpQueue()

    def hasQueuedMessages(self) -> bool:
        return self.steeringQueue.has_items() or self.followUpQueue.has_items()

    @property
    def signal(self) -> AbortSignal | None:
        if self.active_run:
            return self.active_run["abort_controller"].signal
        return None

    def abort(self) -> None:
        if self.active_run:
            self.active_run["abort_controller"].abort()

    async def waitForIdle(self) -> None:
        if self.active_run:
            await self.active_run["future"]

    def reset(self) -> None:
        self._state.messages = []
        self._state._is_streaming = False
        self._state._streaming_message = None
        self._state._pending_tool_calls = set()
        self._state._error_message = None
        self.clearAllQueues()

    async def prompt(
        self,
        input_val: Union[str, AgentMessage, List[AgentMessage]],
        images: List[ImageContent] | None = None,
    ) -> None:
        if self.active_run:
            raise RuntimeError(
                "Agent is already processing a prompt. Use steer() or followUp() to queue messages, or wait for completion."
            )
        messages = self.normalizePromptInput(input_val, images)
        await self.runPromptMessages(messages)

    async def continue_run(self) -> None:
        # Note: continue is a python keyword, renamed to continue_run
        if self.active_run:
            raise RuntimeError(
                "Agent is already processing. Wait for completion before continuing."
            )

        if not self._state.messages:
            raise ValueError("No messages to continue from")

        last_message = self._state.messages[-1]
        if last_message.get("role") == "assistant":
            queued_steering = self.steeringQueue.drain()
            if len(queued_steering) > 0:
                await self.runPromptMessages(queued_steering, skip_initial_steering_poll=True)
                return

            queued_followups = self.followUpQueue.drain()
            if len(queued_followups) > 0:
                await self.runPromptMessages(queued_followups)
                return

            raise ValueError("Cannot continue from message role: assistant")

        await self.runContinuation()

    def normalizePromptInput(
        self,
        input_val: Union[str, AgentMessage, List[AgentMessage]],
        images: List[ImageContent] | None = None,
    ) -> List[AgentMessage]:
        if isinstance(input_val, list):
            return input_val

        if isinstance(input_val, dict):
            return [input_val]

        content: List[Union[TextContent, ImageContent]] = [{"type": "text", "text": input_val}]
        if images:
            content.extend(images)

        return [
            {
                "role": "user",
                "content": content,
                "timestamp": int(time.time() * 1000),
            }
        ]

    async def runPromptMessages(
        self, messages: List[AgentMessage], skip_initial_steering_poll: bool = False
    ) -> None:
        async def execute(signal: AbortSignal) -> None:
            await run_agent_loop(
                messages,
                self.createContextSnapshot(),
                self.createLoopConfig(skip_initial_steering_poll),
                self.processEvents,
                signal,
                self.streamFn,
            )

        await self.runWithLifecycle(execute)

    async def runContinuation(self) -> None:
        async def execute(signal: AbortSignal) -> None:
            await run_agent_loop_continue(
                self.createContextSnapshot(),
                self.createLoopConfig(),
                self.processEvents,
                signal,
                self.streamFn,
            )

        await self.runWithLifecycle(execute)

    def createContextSnapshot(self) -> AgentContext:
        return {
            "systemPrompt": self._state.systemPrompt,
            "messages": list(self._state.messages),
            "tools": list(self._state.tools),
        }

    def createLoopConfig(self, skip_initial_steering_poll: bool = False) -> AgentLoopConfig:
        skip = skip_initial_steering_poll

        async def get_steering() -> List[AgentMessage]:
            nonlocal skip
            if skip:
                skip = False
                return []
            return self.steeringQueue.drain()

        async def get_followup() -> List[AgentMessage]:
            return self.followUpQueue.drain()

        config: AgentLoopConfig = {
            "model": self._state.model,
            "transport": self.transport,
            "toolExecution": self.toolExecution,
            "convertToLlm": self.convertToLlm,
            "getSteeringMessages": get_steering,
            "getFollowUpMessages": get_followup,
        }

        if self.sessionId is not None:
            config["sessionId"] = self.sessionId
        if self.onPayload is not None:
            config["onPayload"] = self.onPayload
        if self.onResponse is not None:
            config["onResponse"] = self.onResponse
        if self.thinkingBudgets is not None:
            config["thinkingBudgets"] = cast(dict[str, int], self.thinkingBudgets)
        if self.maxRetryDelayMs is not None:
            config["maxRetryDelayMs"] = self.maxRetryDelayMs
        if self.beforeToolCall is not None:
            config["beforeToolCall"] = self.beforeToolCall
        if self.afterToolCall is not None:
            config["afterToolCall"] = self.afterToolCall
        if self.transformContext is not None:
            config["transformContext"] = self.transformContext
        if self.getApiKey is not None:
            config["getApiKey"] = self.getApiKey

        if self._state.thinkingLevel != "off":
            config["reasoning"] = self._state.thinkingLevel

        prep_next = self.prepareNextTurn
        if prep_next is not None:

            async def prepare(context: Any) -> Any:
                return await maybe_await(prep_next(self.signal))

            config["prepareNextTurn"] = prepare

        return config

    async def runWithLifecycle(
        self, executor: Callable[[AbortSignal], Coroutine[Any, Any, None]]
    ) -> None:
        if self.active_run:
            raise RuntimeError("Agent is already processing.")

        abort_controller = AbortController()
        loop = asyncio.get_running_loop()
        future = loop.create_future()

        self.active_run = {
            "future": future,
            "abort_controller": abort_controller,
        }

        self._state._is_streaming = True
        self._state._streaming_message = None
        self._state._error_message = None

        try:
            await executor(abort_controller.signal)
        except Exception as error:
            await self.handleRunFailure(error, abort_controller.signal.aborted)
        finally:
            self.finishRun()

    async def handleRunFailure(self, error: Any, aborted: bool) -> None:
        failure_message: AgentMessage = {
            "role": "assistant",
            "content": [{"type": "text", "text": ""}],
            "api": self._state.model.get("api", "unknown"),
            "provider": self._state.model.get("provider", "unknown"),
            "model": self._state.model.get("id", "unknown"),
            "usage": EMPTY_USAGE,
            "stopReason": "aborted" if aborted else "error",
            "errorMessage": str(error),
            "timestamp": int(time.time() * 1000),
        }
        await maybe_await(self.processEvents({"type": "message_start", "message": failure_message}))
        await maybe_await(self.processEvents({"type": "message_end", "message": failure_message}))
        await maybe_await(
            self.processEvents({"type": "turn_end", "message": failure_message, "toolResults": []})
        )
        await maybe_await(self.processEvents({"type": "agent_end", "messages": [failure_message]}))

    def finishRun(self) -> None:
        self._state._is_streaming = False
        self._state._streaming_message = None
        self._state._pending_tool_calls = set()
        if self.active_run:
            if not self.active_run["future"].done():
                self.active_run["future"].set_result(None)
        self.active_run = None

    async def processEvents(self, event: AgentEvent) -> None:
        event_type = event["type"]
        dict_event = cast(dict[str, Any], event)

        if event_type == "message_start":
            self._state._streaming_message = dict_event.get("message")
        elif event_type == "message_update":
            self._state._streaming_message = dict_event.get("message")
        elif event_type == "message_end":
            self._state._streaming_message = None
            msg = dict_event.get("message")
            if msg:
                self._state._messages.append(msg)
        elif event_type == "tool_execution_start":
            tool_call_id = dict_event.get("toolCallId")
            if tool_call_id:
                self._state._pending_tool_calls.add(tool_call_id)
        elif event_type == "tool_execution_end":
            tool_call_id = dict_event.get("toolCallId")
            if tool_call_id:
                self._state._pending_tool_calls.discard(tool_call_id)
        elif event_type == "turn_end":
            msg = dict_event.get("message")
            if msg and msg.get("role") == "assistant" and msg.get("errorMessage"):
                self._state._error_message = msg["errorMessage"]
        elif event_type == "agent_end":
            self._state._streaming_message = None

        signal = self.signal
        if not signal:
            raise RuntimeError("Agent listener invoked outside active run")

        listeners = list(self.listeners)
        if listeners:
            await asyncio.gather(*(maybe_await(listener(event, signal)) for listener in listeners))
